From 56f985a93f2e3555ea9cf052d5c22e54f8c26591 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markus=20M=C3=BCtzel?= <markus.muetzel@gmx.de>
Date: Fri, 28 Aug 2026 13:57:59 +0200
Subject: [PATCH] FindMPI: Parse arguments that are surrounded by single-quotes

Some MPI compiler wrappers (e.g., the ones in MSYS2) surround arguments with
single-quotes (to deal, e.g., with spaces and double-quotes in arguments).

Parse the output of the compiler wrappers taking optional single-quotes into
account.
---
 Modules/FindMPI.cmake | 35 ++++++++++++++++++++++++++++++-----
 1 file changed, 30 insertions(+), 5 deletions(-)

diff --git a/Modules/FindMPI.cmake b/Modules/FindMPI.cmake
index ebfd79fe3e..5c12ec2c01 100644
--- a/Modules/FindMPI.cmake
+++ b/Modules/FindMPI.cmake
@@ -834,9 +834,14 @@ function (_MPI_interrogate_compiler LANG)
   endif()
 
   # Same deal as above, for the definitions.
-  string(REGEX MATCHALL "(^| )${_MPI_PREPROCESSOR_FLAG_REGEX}-D *([^\"\n ]+|\"[^\"]+\")" MPI_ALL_COMPILE_DEFINITIONS "${MPI_COMPILE_CMDLINE}")
+  string(REGEX MATCHALL "(^| )'?${_MPI_PREPROCESSOR_FLAG_REGEX}-D *([^\"\n ]+|\"[^\"]+\"|'[^']')'?" MPI_ALL_COMPILE_DEFINITIONS "${MPI_COMPILE_CMDLINE}")
 
   foreach(_MPI_COMPILE_DEFINITION IN LISTS MPI_ALL_COMPILE_DEFINITIONS)
+    # strip single-quotes around arguments
+    string(REGEX MATCH "^ ?'([^']*)'$" _t "${_MPI_COMPILE_DEFINITION}")
+    if(CMAKE_MATCH_1)
+      set(_MPI_COMPILE_DEFINITION ${CMAKE_MATCH_1})
+    endif()
     string(REGEX REPLACE "^ ?${_MPI_PREPROCESSOR_FLAG_REGEX}-D *" "" _MPI_COMPILE_DEFINITION "${_MPI_COMPILE_DEFINITION}")
     string(REPLACE "\"" "" _MPI_COMPILE_DEFINITION "${_MPI_COMPILE_DEFINITION}")
     if(NOT _MPI_COMPILE_DEFINITION MATCHES "^_FORTIFY_SOURCE.*")
@@ -845,7 +850,7 @@ function (_MPI_interrogate_compiler LANG)
   endforeach()
 
   # Extract include paths from compile command line
-  string(REGEX MATCHALL "(^|\n| )${_MPI_PREPROCESSOR_FLAG_REGEX}${CMAKE_INCLUDE_FLAG_${LANG}} *([^\"\n ]+|\"[^\"]+\")"
+  string(REGEX MATCHALL "(^|\n| )'?${_MPI_PREPROCESSOR_FLAG_REGEX}${CMAKE_INCLUDE_FLAG_${LANG}} *([^\"\n ]+|\"[^\"]+\"|'[^']')'?"
     MPI_ALL_INCLUDE_PATHS "${MPI_COMPILE_CMDLINE}")
 
   # If extracting failed to work, we'll try using -showme:incdirs.
@@ -859,6 +864,11 @@ function (_MPI_interrogate_compiler LANG)
   endif()
 
   foreach(_MPI_INCLUDE_PATH IN LISTS MPI_ALL_INCLUDE_PATHS)
+    # strip single-quotes around arguments
+    string(REGEX MATCH "^ ?'([^']*)'$" _t "${_MPI_INCLUDE_PATH}")
+    if(CMAKE_MATCH_1)
+      set(_MPI_INCLUDE_PATH ${CMAKE_MATCH_1})
+    endif()
     string(REGEX REPLACE "^ ?${_MPI_PREPROCESSOR_FLAG_REGEX}${CMAKE_INCLUDE_FLAG_${LANG}} *" "" _MPI_INCLUDE_PATH "${_MPI_INCLUDE_PATH}")
     string(REPLACE "\n" "" _MPI_INCLUDE_PATH "${_MPI_INCLUDE_PATH}")
     string(REPLACE "\"" "" _MPI_INCLUDE_PATH "${_MPI_INCLUDE_PATH}")
@@ -868,9 +878,14 @@ function (_MPI_interrogate_compiler LANG)
   endforeach()
 
   # The next step are linker flags and library directories. Here, we first take the flags given in raw -L or -LIBPATH: syntax.
-  string(REGEX MATCHALL "(^| )${CMAKE_LIBRARY_PATH_FLAG} *([^\"\n ]+|\"[^\"]+\")" MPI_DIRECT_LINK_PATHS "${MPI_LINK_CMDLINE}")
+  string(REGEX MATCHALL "(^| )'?${CMAKE_LIBRARY_PATH_FLAG} *([^\"'\n ]+|\"[^\"]+\"|'[^']')'?" MPI_DIRECT_LINK_PATHS "${MPI_LINK_CMDLINE}")
   foreach(_MPI_LPATH IN LISTS MPI_DIRECT_LINK_PATHS)
-    string(REGEX REPLACE "(^| )${CMAKE_LIBRARY_PATH_FLAG} *" "" _MPI_LPATH "${_MPI_LPATH}")
+    # strip single-quotes around arguments
+    string(REGEX MATCH "^ ?'([^']*)'$" _t "${_MPI_LPATH}")
+    if(CMAKE_MATCH_1)
+      set(_MPI_LPATH ${CMAKE_MATCH_1})
+    endif()
+    string(REGEX REPLACE "^ ?${CMAKE_LIBRARY_PATH_FLAG} *" "" _MPI_LPATH "${_MPI_LPATH}")
     list(APPEND MPI_ALL_LINK_PATHS "${_MPI_LPATH}")
   endforeach()
 
@@ -930,6 +945,11 @@ function (_MPI_interrogate_compiler LANG)
 
   # We need to remove potential quotes and convert the paths to CMake syntax while resolving them, too.
   foreach(_MPI_LPATH IN LISTS MPI_ALL_LINK_PATHS)
+    # strip single-quotes around arguments
+    string(REGEX MATCH "^ ?'([^']*)'$" _t "${_MPI_LPATH}")
+    if(CMAKE_MATCH_1)
+      set(_MPI_LPATH ${CMAKE_MATCH_1})
+    endif()
     string(REPLACE "\"" "" _MPI_LPATH "${_MPI_LPATH}")
     get_filename_component(_MPI_LPATH "${_MPI_LPATH}" REALPATH)
     list(APPEND MPI_LINK_DIRECTORIES_WORK "${_MPI_LPATH}")
@@ -938,10 +958,15 @@ function (_MPI_interrogate_compiler LANG)
   # Extract the set of libraries to link against from the link command line
   # This only makes sense if CMAKE_LINK_LIBRARY_FLAG is defined, i.e. a -lxxxx syntax is supported by the compiler.
   if(CMAKE_LINK_LIBRARY_FLAG)
-    string(REGEX MATCHALL "(^| )${CMAKE_LINK_LIBRARY_FLAG}([^\"\n ]+|\"[^\"]+\")"
+    string(REGEX MATCHALL "(^| )'?${CMAKE_LINK_LIBRARY_FLAG}([^\"\n ]+|\"[^\"]+\"|'[^']')'?"
       MPI_LIBNAMES "${MPI_LINK_CMDLINE}")
 
     foreach(_MPI_LIB_NAME IN LISTS MPI_LIBNAMES)
+      # strip single-quotes around arguments
+      string(REGEX MATCH "^ ?'([^']*)'$" _t "${_MPI_LIB_NAME}")
+      if(CMAKE_MATCH_1)
+        set(_MPI_LIB_NAME ${CMAKE_MATCH_1})
+      endif()
       # also match flags starting with "-l:" here
       string(REGEX REPLACE "^ ?${CMAKE_LINK_LIBRARY_FLAG}(:lib|:)?" "" _MPI_LIB_NAME "${_MPI_LIB_NAME}")
       string(REPLACE "\"" "" _MPI_LIB_NAME "${_MPI_LIB_NAME}")
-- 
2.51.0.windows.2

